virtual_rainforest/core/config.py

```diff
@@ -236,6 +233,80 @@ def load_in_config_files(files: list[Path]) -> dict:
     return config_dict
 
 
+def extend_with_default(
+    validator_class: type[Draft202012Validator],
+) -> type[Draft202012Validator]:
+    """Extend validator so that it can populate default values from the schema.
+
+    Args:
+        validator_class: Validator to be extended
+    """
+
+    validate_properties = validator_class.VALIDATORS["properties"]
+
+    def set_defaults(
+        validator: type[Draft202012Validator],
+        properties: dict,
+        instance: dict,
+        schema: dict,
+    ) -> Iterator:
+        """Generate an iterator to populate defaults."""
+        for property, subschema in properties.items():
+            if "default" in subschema:
+                instance.setdefault(property, subschema["default"])
+
+        for error in validate_properties(
+            validator,
+            properties,
+            instance,
+            schema,
+        ):
+            yield error
+
+    return validators.extend(
+        validator_class,
+        {"properties": set_defaults},
+    )
+
+
+def add_core_defaults(config_dict: dict) -> dict:
+    """Add default config options for the core module to the config dictionary.
+
+    This is a separate function because the default modules to load are specified in the
+    core schema. So, these defaults must be populated before the complete schema can be
+    constructed.
+
+    Args:
+        config_dict: The complete configuration settings for the particular model
+            instance
+    Raises:
+        RuntimeError: If the core module schema can't be found, or if it cannot be
+            validated
+    """
+
+    # Make a new validator that allows the addition of defaults
+    ValidatorWithDefaults = extend_with_default(Draft202012Validator)
+
+    # Look for core config in schema registry
+    if "core" in SCHEMA_REGISTRY:
+        core_schema = SCHEMA_REGISTRY["core"]
+    else:
+        log_and_raise(
+            "Expected a schema for core module configuration, it was not provided!",
+            RuntimeError,
+        )
+
+    try:
+        ValidatorWithDefaults(core_schema).validate(config_dict)
+    except exceptions.ValidationError as err:
+        log_and_raise(
+            f"Validation of core configuration files failed: {err.message}",
+            RuntimeError,
```

Comment:
I'd personally consider a custom exception type for this. It seems like there are several places where configuration-related exceptions can be raised and it'd be useful to let the caller check for this specifically.